keps/sig-cli/20190318-kustomize-generators-transformers.md

```diff
@@ -0,0 +1,389 @@
+to its set of Resources.
+
+#### Bespoke Abstractions for Organizations
+
+Bob's organization deploys many variants of the same high-level conceptual set of Resources.  Creating the Resource Config
+for each instance that needs to be deployed requires lots of boilerplate.  Instead Bob's organization develops
+tools for generating standardized Resource Config based off of some inputs.
+
+Bob creates a new generator for his organization that allow higher level abstractions to be defined as
+new virtual Resource types that don't exist in the cluster, but are used to generate low-level types.
+
+1. Bob builds and installs the new generator plugin `$XDG_CONFIG_HOME/kustomize/plugins/generators/team.example.com`.
+1. Bob creates the `app.yaml` Resource Config and adds it to his `kustomization.yaml`
+
+```yaml
+groupVersion: team.example.com/v1beta1
+kind: CommonApp
+generate:
+  image: foo
+  size: big
+```
+
+```yaml
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: Kustomization
+
+generators:
+- app.yaml
+```
+
+The `team.example.com` binary is invoked with the `app.yaml` contents passed on STDIN.  Kustomize adds the
+emitted Resource to its set of Resources.
+
+If multiple generators are specified, they are run in the order they are listed.
+
+#### Bespoke Configuration for Organizations
+
+Alice's Organization requires that various fields are defaulted if unset.  SRE would like to be able to see
+the full Resources that are being Applied and have this auditable in an scm, such as git, rather than
+having Webhooks provide server-side mutions that are not capture in review or scm.
+
+1. Alice builds and installs the new transformer plugin at `$XDG_CONFIG_HOME/kustomize/plugins/transformers/team.example.com`
+1. Alice creates the `transform.yaml` Resource Config and adds it to her `kustomization.yaml` as a `transformers` entry.
+
+```yaml
+groupVersion: team.example.com
+kind: CommonApp
+transform:
+  namePrefix: foo-
+  commonAnnotations:
+    foo: bar
+```
+
+```yaml
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: Kustomization
+
+transformers:
+- transform.yaml
+```
+
+The `team.example.com` binary is invoked with both the `transform.yaml` contents and the Resources passed on STDIN.
+Kustomize replaces its Resources with the emitted Resources.
+
+### Implementation Details/Notes/Constraints [optional]
+
+### Risks and Mitigations
+
+## Design Details
+
+### Test Plan
+
+
+### Graduation Criteria
+
+TBD
+
+Consider:
+
+- Executing plugins for generated Resources
+
+### Upgrade / Downgrade Strategy
+
+NA - Client side only
+
+### Version Skew Strategy
+
+NA - Client side only
+
+## Implementation History
+
+
+## Drawbacks [optional]
+
+It allows users to do very complex thing with Kustomize.
+
+## Alternatives [optional]
+
+- Imperatively piping generator executables to kubectl apply
+- Writing scripts to invoke generator executables and piping them to kubectl apply
```

Comment:
- Start a new "meta-kustomize" tool that implements all this plugin machinery - and then make the existing kustomize just one of the plugins within that (rather than having the meta version burst from the stomach of the existing code and consume the host - yay aliens metaphors)

Reply:
This is listed under non-goals - re-implemented kustomize as plugins.

Reply:
yes, and that's very reasonable.  I'm suggesting (for completeness, not a serious proposal) that we implement _this_ KEP in a new tool/layer (that isn't existing-kustomize).

Reply:
Ah got it.  Yeah I can add that.